core/ipc/rpc/kafka/src/main/java/org/opennms/core/ipc/rpc/kafka/KafkaRpcClientFactory.java

```diff
@@ -127,34 +146,64 @@
                 delayQueue.offer(responseHandler);
                 rpcResponseMap.put(rpcId, responseHandler);
                 kafkaConsumerRunner.startConsumingForModule(module.getId());
-                if (request.getSystemId() != null) {
+                byte[] messageInBytes = marshalRequest.getBytes();
+                int totalChunks = IntMath.divide(messageInBytes.length, MAX_BUFFER_SIZE, RoundingMode.UP);
+                // Divide the message in chunks and send each chunk as a different message with the same key.
+                for (int chunk = 0; chunk < totalChunks; chunk++) {
+                    // Calculate remaining bufferSize for each chunk.
+                    int bufferSize = KafkaRpcConstants.getBufferSize(messageInBytes.length, MAX_BUFFER_SIZE, chunk);
+                    ByteString byteString = ByteString.copyFrom(messageInBytes, chunk * MAX_BUFFER_SIZE, bufferSize);
                     // For directed RPCs, send request to all partitions (consumers),
-                    // this forces partitions >= max(number of minions at location)
-                    RpcMessageProtos.RpcMessage rpcMessage = RpcMessageProtos.RpcMessage.newBuilder()
-                                                                 .setRpcId(rpcId)
-                                                                 .setSystemId(request.getSystemId())
-                                                                 .setExpirationTime(expirationTime)
-                                                                 .setRpcContent(ByteString.copyFromUtf8(marshalledRequest))
-                                                                 .build();
-                    List<PartitionInfo> partitionInfo = producer.partitionsFor(requestTopic);
-                    partitionInfo.stream().forEach(partition -> {
+                    // as it is reasonable to assume that partitions >= consumers(number of minions at location)
+                    int chunkNum = chunk;
+                    if (request.getSystemId() != null) {
+                        RpcMessageProtos.RpcMessage rpcMessage = RpcMessageProtos.RpcMessage.newBuilder()
```

Comment:
I had made a comment about this builder but somehow it got lost in my original review. I mentioned for brevity we could create one builder and re-use it rather than a new one for each loop iteration. This would save on line count inside the loop and also save on # objects created.

So outside the loop we could have:
```
RpcMessageProtos.RpcMessage.Builder rpcMessageBuilder = RpcMessageProtos.RpcMessage.newBuilder()
                        .setRpcId(rpcId)
                        .setExpirationTime(expirationTime)
                        .setTotalChunks(totalChunks);
```

Just inside the loop:
```
rpcMessageBuilder.setRpcContent(byteString).setCurrentChunkNumber(chunk);
```

In the conditional:
```
if (request.getSystemId() != null) {
                        RpcMessageProtos.RpcMessage rpcMessage =
                                rpcMessageBuilder.setSystemId(request.getSystemId()).build();
```

and in the else:
```
} else {
                        RpcMessageProtos.RpcMessage rpcMessage =
                                rpcMessageBuilder.setSystemId("").build();
```